### ISSUE TITLE
Run Jobs inside Jobs

### ISSUE THREAD

**Issue by thieman** [thieman/dagobah#60]
It would be cool to be able to run a Job inside of another Job (basically like it's a Task). For those of you not familiar with the insane jargon, I'm talking about a DAG that could look like, for a simple example:

**Job 1**: Task 1 -> Job 2 -> Task 2

**Job 2**: Task 3 -> Task 4

Job 1 would complete Task 1, start Job 2, monitor its completion status, and if it finishes successfully then proceed on to Task 2.

I think the first step is to make Jobs _either_ scheduled or able to be called from inside other Jobs. There will be a lot of edge cases to argue about along the way.

<img src="http://bossip.files.wordpress.com/2012/09/xzibit.png?w=500" width=150></img>

**Comment by rclough** [comment]
so #160 is in, and I'm starting on "expanding" jobs into one giant graph (when the "snapshot" is initialized).

I'm thinking I'll need to add a `predecessors(node)` to py-dag in order to implement this. This is because, when we expand a `JobTask`, we need all of the `JobTask`'s `ind_nodes` to have edges coming from the `JobTask`'s predecessors ("parents", if this were a tree).

Also, I'd want to add another downstream function, `downstreams_final(node)` to get all of the nodes which ONLY don't have further nodes downstream (I assume `downstream_all` gets you all the nodes in between the current and the "final" downstreams as well)

I'm working on code now and it's actually fairly complex (like how does one merge the edge and node dictionaries properly? for example) so any thoughts are welcome.

**Comment by rclough** [comment]
let me revise that `downstreams_final` idea.

What's actually needed is all nodes in the DAG that have no downstreams. Not sure what a good name for this is, or if theres a canonical name.

**Comment by rclough** [comment]
Here's my first pass at coding it. This is untested, and will not run because I havent implemented some of the aforementioned functions (I decided to call that last one `dag_ends()` for the "ends" of a DAG). But you get the gist:

```python
    def expand(self, dag=None):
        """
        This method takes some DAG object, and expands all the underlying
        tasks recursively:
            * Starting with ind_nodes, traverse the graph
            * If the node is expandable:
                * Call expand on the JobTask, save expanded version "Expanded"
                * Connect all parents of this JobTask to the ind_nodes of
                  "Expanded"
                * Get a list of all downstream nodes in "Expanded" that don't
                  have a downstream node, and add edges to all downstream
                  nodes of the JobTask.
            * Delete JobTask from graph
            * Continue traversing from the downstreams of the deleted JobTask
        """
        dag = dag.deepcopy()
        traversal_queue = [dag.tasks[t] for t in dag.ind_nodes()]
        expanded = []
        while traversal_queue:
            task = traversal_queue.pop()
            if self.implements_expandable(task) and task.name not in expanded:
                expanded.append(task.name)
                expanded = task.expand()

                # Empty Job expansion
                if not expanded:
                    pred = dag.predecessors(task.name)
                    children = dag.downstream(task.name)
                    [dag.add_edge(p, c) for p in pred for c in children]
                    dag.delete_task(task.name)
                    continue

                # Prepend all expanded task names with "<jobname>_" to prevent
                # task name conflicts
                [expanded.edit_task(t, {"name": "{0}_{1}".format(task.name, t)})
                 for t in expanded.tasks]

                # Merge node and edge dictionaries (creating 2 unconnected DAGs,
                # in one graph)
                final_dict = default_dict(list)
                for key,value in dag.graph:
                    final_dict[key].append[value]
                for key,value in expanded.graph:
                    final_dict[key].expand(value)
                dag.graph = final_dict

                # Add edges between predecessors and start nodes
                predecessors = dag.predecessors(task.name)
                start_nodes = expanded.ind_nodes()
                [dag.add_edge(p, s) for p in predecessors for s in start_nodes]

                # Add edges between the final downstreams and the child nodes
                final_tasks = expanded.dag_ends()
                children = dag.downstream(task.name)
                [dag.add_edge(f, c) for f in final_tasks for c in children]

                # Delete old reference
                dag.delete_node(task.name)

```

**Comment by rclough** [comment]
Also this is probably obvious, but I did a breadth-first traversal, not sure if there's any advantage over depth first, but its just hat my brain defaulted to

**Comment by rclough** [comment]
maybe it should be called `end_nodes()` which goes nicely with `ind_nodes()` which are basically the opposite concept (nodes with no incoming edges vs nodes with no outgoing edges)

**Comment by thieman** [comment]
> I'm thinking I'll need to add a predecessors(node) to py-dag in order to implement this. This is because, when we expand a JobTask, we need all of the JobTask's ind_nodes to have edges coming from the JobTask's predecessors ("parents", if this were a tree).

This makes sense to me, if I remember the py-dag structure this should be pretty simple and hopefully O(n).

> What's actually needed is all nodes in the DAG that have no downstreams. Not sure what a good name for this is, or if theres a canonical name.

I believe the standard term here is "leaf node." So maybe `all_leaves`?

Would review the actual code right now but my brain isn't having it, will take a stab at it tomorrow.

**Comment by rclough** [comment]
yeah, since the edges are stored in a map of Edge->list of edges, it can be done in O(n) simply by traversing the graph keys once

**Comment by thieman** [comment]
Thoughts after looking at the code:

1. Does the "prepend expanded tasks with <jobname>_" thing allow us to have multiple instances of the same JobTask within the Job? Or do we need to add some sort of counter component to allow that?
1. Does `edit_task` do any database access or is it just an in-place object mutation?
1. I don't really understand what the "merge node and edge dictionaries" step is doing or why we need it.
1. It seems like this could traverse over the same JobTask multiple times if it has many parents, what is accounting for that?

**Comment by rclough** [comment]
1. By restriction of dagobah, we cannot have 2 tasks with the same name. So 2 of the same jobtasks would have to have different TASK names. So by prepending the child tasks with the task names, we ensure no name conflicts (mostly)
2. Good question, I'll need to look into it, but this should be a copy of the dag in any case. good eye though. I'm about to get off work so I'll check soon.
3. This is adding all of the tasks from the underlying jobtask's job to the current job
4. I think I need a bit more clarification on this?

**Comment by rclough** [comment]
RE # 4, I think I get it. The expanded_tasks list prevents re-expanding a task once it has been expanded once

**Comment by rclough** [comment]
so I realized last week that both the underlying DAG (the `job.graph`, in practice) and the task list need to be synced if we want to allow in place task expansion, and basically in this implementation, they are combined. I'm working on an actual PR now that will manage both basically, by adding a "snapshot_tasks" list just like we do with a snapshot of the DAG
